packages/discount-code-importer/src/main.js

```diff
@@ -1,3 +1,221 @@
+/* @flow */
+import type {
+  LoggerOptions,
+  ChunkOptions,
+  CodeDataArray,
+  CodeData,
+  ConstructorOptions,
+  Summary,
+} from 'types/discountCodes'
+import type {
+  Client,
+  AuthMiddlewareOptions,
+  SyncAction,
+ } from 'types/sdk'
+import npmlog from 'npmlog'
+import Promise from 'bluebird'
+import _ from 'lodash'
+import { createClient } from '@commercetools/sdk-client'
+import { createRequestBuilder } from '@commercetools/api-request-builder'
+import { createAuthMiddlewareForClientCredentialsFlow }
+  from '@commercetools/sdk-middleware-auth'
+import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
+import { createUserAgentMiddleware }
+  from '@commercetools/sdk-middleware-user-agent'
+import { createSyncDiscountCodes } from '@commercetools/sync-actions'
+import { version } from '../package.json'
+
+export default class DiscountCodeImport {
+  // Set flowtype annotations
+  batchSize: number;
+  continueOnProblems: boolean;
+  client: Client;
+  apiConfig: AuthMiddlewareOptions;
+  logger: LoggerOptions;
+  _summary: Summary;
+  syncDiscountCodes: SyncAction;
+
+  constructor (
+    logger: LoggerOptions,
+    options: ConstructorOptions,
+  ) {
+    this.batchSize = options.batchSize || 50
+    this.apiConfig = options.apiConfig
+    this.continueOnProblems = options.continueOnProblems || false
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: 'discount-code-importer',
+          libraryVersion: version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.host }),
+      ],
+    })
+
+    this.syncDiscountCodes = createSyncDiscountCodes()
+
+    this.logger = logger || {
+      error: npmlog.error.bind(this, ''),
+      warn: npmlog.warn.bind(this, ''),
+      info: npmlog.info.bind(this, ''),
+      verbose: npmlog.verbose.bind(this, ''),
+    }
+    this._resetSummary()
+  }
+
+  // Use static method because this is not called on any instance
+  static _buildPredicate (codeObjects: CodeDataArray): string {
+    const predicateArray = codeObjects.map(codeObject => codeObject.code)
+    return `code in ("${predicateArray.join('", "')}")`
+  }
+
+  performStream (chunk: ChunkOptions, cb: () => mixed) {
+    this._processBatches(chunk)
+      .then(() => cb())
+      .catch(err => cb(err.body))
+  }
+
+  _processBatches (codes: CodeDataArray) {
+    // Batch to `batchSize` to reduce necessary fetch API calls
+    const batchedList = _.chunk(codes, this.batchSize)
+    return Promise.map(batchedList, (codeObjects: CodeDataArray) => {
+      // Build predicate and fetch existing code
+      const predicate = DiscountCodeImport._buildPredicate(codeObjects)
+      const service = createRequestBuilder({
+        projectKey: this.apiConfig.projectKey,
+      })
+      const uri = service.discountCodes
+        .where(predicate)
+        .perPage(this.batchSize)
```

Comment:
There can't be more items matching it because the predicate also has the same amount of codes and the codes are unique (there could only be less if the code doesn't yet exist)